### PR TITLE
[GraphBolt] Refactor csc format sampled subgraph.

### PR DESCRIPTION
**Pull request by yxy235**
## Description

This is still a draft of csc format sampled subgraph.
Some functions have been added **temporary** variable so that sampled subgraph **with deduplication** can run successfully, which be delete when everything is ready.
For there is still some misunderstanding about `type_per_edge`, temporarily use `mask` instead of binary search to construct indptr and indices in the subgraph. The final solution will be determined after more detailed research and discussions.

### Homo Sampled Subgraph
```
SampledSubgraphImpl{
        node_pairs: {
                indptr: torch.tensor(),
                indice: torch.tensor(),
        },
        original_column_node_ids: torch.tensor(),
        original_row_node_ids: torch.tensor(),
        original_edge_ids: torch.tensor(),
}
```
### Hetero Sampled Subgraph
```
SampledSubgraphImpl{
        node_pairs: {
                "n0:e0:n1":{
                        indptr: torch.tensor(),
                        indice: torch.tensor(),
                },
        },
        original_column_node_ids: torch.tensor(),
        original_row_node_ids: torch.tensor(),
        original_edge_ids: torch.tensor(),
}
```
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/impl/fused_csc_sampling_graph.py [hunk 1]
```diff
@@ -364,13 +367,110 @@ def _convert_to_homogeneous_nodes(self, nodes):
             homogeneous_nodes.append(ids + self.node_type_offset[ntype_id])
         return torch.cat(homogeneous_nodes)
 
+    def _new_convert_to_sampled_subgraph(
+        self,
+        C_sampled_subgraph: torch.ScriptObject,
+    ):
```
Comment (Skeleton003) [inline python/dgl/graphbolt/impl/fused_csc_sampling_graph.py:373]:
```suggestion
    ) -> newFusedSampledSubgraphImpl:
```

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 2]
```diff
@@ -67,3 +67,64 @@ def __post_init__(self):
...
+    {"A:relation:B": {"indptr": torch.tensor([0, 1, 2, 3]), "indice": torch.tensor([0, 1, 2])}}
+    >>> print(subgraph.original_column_node_ids)
+    {'B': tensor([10, 11, 12])}
+    >>> print(subgraph.original_row_node_ids)
+    {'A': tensor([13, 14, 15])}
+    >>> print(subgraph.original_edge_ids)
+    {"A:relation:B": tensor([19, 20, 21])}
+    """
+    node_pairs: Union[
+        Dict[str, torch.Tensor],
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:99]: Using dict tensor is probably not a good idea.
Consider define a python data class.

@dataclass
class CSCFormatBase:

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 3]
```diff
@@ -67,3 +67,64 @@ def __post_init__(self):
             assert all(
                 isinstance(item, torch.Tensor) for item in self.node_pairs
             ), "Nodes in pairs should be of type torch.Tensor."
+
+
+@dataclass
+class newFusedSampledSubgraphImpl(SampledSubgraph):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:73]: SampledSubgraphImpl

File: python/dgl/graphbolt/impl/fused_csc_sampling_graph.py [hunk 4]
```diff
@@ -364,13 +367,110 @@ def _convert_to_homogeneous_nodes(self, nodes):
...
+            for (i, seed) in enumerate(column):
+                l = indptr[i].item()
+                r = indptr[i + 1].item()
+                node_type = _get_nodetype(seed, self.node_type_offset)
+                for (etype, etype_id) in node_edge_type[node_type]:
+                    src_ntype, _, _ = etype_str_to_tuple(etype)
+                    src_ntype_id = self.metadata.node_type_to_id[src_ntype]
+                    mask = type_per_edge[l:r] == etype_id
+                    sub_indice = indices[l:r][mask]
+                    # [TODO] These comments in this loop and commented-out
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/fused_csc_sampling_graph.py:431]: This is a little bit confusing, remove it or make it easier for people to understand the future work.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 5]
```diff
@@ -67,3 +67,83 @@ def __post_init__(self):
...
+    {"A:relation:B": CSCForamtBase(indptr=torch.tensor([0, 1, 2, 3]),
+    ... indices=torch.tensor([0, 1, 2]))}
+    >>> print(subgraph.original_column_node_ids)
+    {'B': tensor([10, 11, 12])}
+    >>> print(subgraph.original_row_node_ids)
+    {'A': tensor([13, 14, 15])}
+    >>> print(subgraph.original_edge_ids)
+    {"A:relation:B": tensor([19, 20, 21])}
+    """
+    node_pairs: Union[
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:117]: Change the name as it is not `node_pairs `.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 6]
```diff
@@ -67,3 +67,83 @@ def __post_init__(self):
             assert all(
                 isinstance(item, torch.Tensor) for item in self.node_pairs
             ), "Nodes in pairs should be of type torch.Tensor."
+
+
+@dataclass
+class CSCFormatBase:
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:73]: Move this to `base.py` as it is likely to be used by others.

File: python/dgl/graphbolt/impl/fused_csc_sampling_graph.py [hunk 7]
```diff
@@ -923,3 +1040,27 @@ def from_dglgraph(
...
+    ntype = -1
+    for offset in node_type_offset:
+        if index >= offset:
+            ntype += 1
+        else:
+            break
+    return ntype
+
+
+def _binary_search_etype_end(type_per_edge, l, r, etype_id):
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/fused_csc_sampling_graph.py:1056]: See https://pytorch.org/docs/stable/generated/torch.searchsorted.html. but I'm not sure it 100% applied to this case.

File: python/dgl/graphbolt/impl/fused_csc_sampling_graph.py [hunk 8]
```diff
@@ -923,3 +1040,27 @@ def from_dglgraph(
         ),
         metadata,
     )
+
+
+def _get_nodetype(index, node_type_offset):
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/fused_csc_sampling_graph.py:1045]: Also use binary search.

File: python/dgl/graphbolt/impl/fused_csc_sampling_graph.py [hunk 9]
```diff
@@ -440,8 +551,10 @@ def sample_neighbors(
         C_sampled_subgraph = self._sample_neighbors(
             nodes, fanouts, replace, probs_name
         )
-
-        return self._convert_to_sampled_subgraph(C_sampled_subgraph)
+        if deduplicate is True:
```
Comment (peizhou001) [inline python/dgl/graphbolt/impl/fused_csc_sampling_graph.py:554]: why not `_convert_to_sampled_subgraph` and `_convert_to_fused_sampled_subgraph`

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 5 above)
Comment (yxy235) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:99]: For the name is defined in `Sampled Subgraph` and it is also used by `Fused Sampled Subgraph Impl`, it is hard to change the name in `Sampled subgraph impl`. Should I change the name in the Sampled Subgraph or what?

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 10]
```diff
@@ -67,3 +67,65 @@ def __post_init__(self):
...
             ), "Nodes in pairs should be of type torch.Tensor."
+
+
+@dataclass
+class SampledSubgraphImpl(SampledSubgraph):
+    r"""Sampled subgraph of CSCSamplingGraph.
+
+    Examples
+    --------
+    >>> node_pairs = {"A:relation:B": CSCFormatBase(indptr=torch.tensor([0, 1, 2, 3]),
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:78]: make sure all the node_pairs related to CSCFormatBase are renamed to csc_format.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 5 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:99]: Ah, I see, good call, let's leave it as it, and refactor it later.

**Review by frozenbugs — APPROVED** [review]
